### ISSUE TITLE
Occasional segmentation fault when non-blocking function is interrupted by end of script

### ISSUE THREAD

**Issue by mgeier** [spatialaudio/python-sounddevice#1]
This only happens if a script ends before playback is finished.

This isn't a real problem since you would use a blocking call or at least `sd.wait()` in a Python script.
But crashing the interpreter isn't really nice ...

This is an example script to reproduce the error:

``` Python
import sounddevice as sd
import numpy as np

sd.play(np.zeros(100000))
```

Note that the crash doesn't happen every time, only occasionally.

**Comment by tgarc** [comment]
I was trying to look at this but I quickly realized that sounddevice isn't actually a compiled extension which makes it more difficult to debug C issues like this. Is it much more troublesome to make it into a compiled extension?

**Comment by mgeier** [comment]
I'm not sure if I understand ...

It would be reasonably simple to switch from CFFI's ABI mode to API mode, see http://cffi.readthedocs.io/en/latest/cdef.html.
This would change hardly anything in the implementation, most notably, the callback would still be implemented in Python.

If by "compiled extension" you mean manually writing C code using the Python C API, this would be an insane amount of work. Also, PyAudio already does that: https://people.csail.mit.edu/hubert/pyaudio/.
In this case, the callback would still be written in Python.

If you mean to keep CFFI (in API mode) but implement the callback in C, that's what I'm doing in https://github.com/mgeier/python-rtmixer/.

Finally, if you are talking about using the Python C API *and* implementing the callback in C, that's what pyo does: https://github.com/belangeo/pyo.

**Comment by tgarc** [comment]
Sorry about the vagueness, I actually just meant using CFFI in API mode. All this would do is produce some debuggable C code.

**Comment by mgeier** [comment]
OK. It should be straightforward, but I haven't actually tried it.

Do you want me to make a branch for API mode?

**Comment by tgarc** [comment]
Nah, I wouldn't put this at a high priority. I will experiment with it in
the near future. I've done it for a previous project and I agree it
*should* be straightforward.

**Comment by mgeier** [comment]
Fixed by #75.
